nixos/modules/services/monitoring/osquery.nix

```
@@ -0,0 +1,114 @@
+
+  osqueryi = pkgs.runCommand "osqueryi" { nativeBuildInputs = [ pkgs.makeWrapper ]; } ''
+    mkdir -p $out/bin
+    ln -s ${pkgs.osquery}/bin/osqueryi $out/bin/osqueryi
+    wrapProgram $out/bin/osqueryi \
+      --add-flags "--flagfile ${flagfile}"
+    '';
+in
+{
+  options.services.osquery = {
+    enable = mkEnableOption (mdDoc "osqueryd daemon");
+
+    settings = mkOption {
+      default = { };
+      description = mdDoc ''
+        Configuration to be written to the osqueryd JSON configuration file.
+        To understand the configuration format, refer to https://osquery.readthedocs.io/en/stable/deployment/configuration/#configuration-components.
+      '';
+      example = {
+        options.utc = false;
+      };
+      type = types.attrs;
+    };
+
+    flags = mkOption {
+      default = { };
+      description = mdDoc ''
+        Attribute set of flag names and values to be written to the osqueryd flagfile.
+        For more information, refer to https://osquery.readthedocs.io/en/stable/installation/cli-flags.
+      '';
+      example = {
+        config_refresh = "10";
+      };
+      type = with types;
+        let
+          pathWithPrefix = prefix: mkOptionType {
+            name = "path with prefix";
+            description = "path with prefix \"${prefix}\"";
+            descriptionClass = "noun";
+            check = with (lib.strings);
+              x: path.check x && hasPrefix prefix (toString x);
+            merge = mergeEqualOption;
+          };
+        in
+        submodule {
+          freeformType = attrsOf str;
+          options = {
+            database_path = mkOption {
+              default = stateDirectoryPrefix + "osquery/osquery.db";
+              description = mdDoc "Path used for the database file.";
+              type = pathWithPrefix stateDirectoryPrefix;
+            };
+            logger_path = mkOption {
+              default = logsDirectoryPrefix + "osquery";
+              description = mdDoc "Base directory used for logging.";
+              # Systemd sytem unit LogsDirectory starts with /var/log/.
+              type = pathWithPrefix logsDirectoryPrefix;
+            };
+            pidfile = mkOption {
+              default = "/run/osquery/osqueryd.pid";
+              description = mdDoc "Path used for pid file.";
+              type = types.path;
+            };
+          };
+        };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ osqueryi ];
+    systemd.services.osqueryd = {
+      after = [ "network.target" "syslog.service" ];
+      description = "The osquery daemon";
+      serviceConfig = with lib.strings; {
+        ExecStart = "${pkgs.osquery}/bin/osqueryd --flagfile ${flagfile}";
+        PIDFile = cfg.flags.pidfile;
+
+        LogsDirectory = [ (removePrefix logsDirectoryPrefix cfg.flags.logger_path) ];
+        StateDirectory = [
+            (removePrefix stateDirectoryPrefix (dirname cfg.flags.database_path))
+        ];
```

```suggestion
        LogsDirectory = removePrefix logsDirectoryPrefix cfg.flags.logger_path;
        StateDirectory = stateDirectoryPrefix (dirname cfg.flags.database_path);
```
we don't need lists if we don't overwrite things and then we need and empty entry first.

---

Correct. Done.